stan/math/prim/mat/fun/append_array.hpp

```diff
@@ -1,80 +1,141 @@
+     * is recursively called on each element of x to copy it to an element of y.
      *
-     * @tparam T1 Type of element in Matrices in first vector
-     * @tparam T2 Type of element in Matrices in second vector
-     * @tparam R Row specification of matrices
-     * @tparam C Column specification of matrices
-     * @param x First vector
-     * @param y Second vector
-     * @return A vector of x and y concatenated together (in that order)
+     * @tparam Type of source
+     * @tparam Type of destination
+     * @param x Source variable
+     * @param y Target variable
      */
-    template <typename T1, typename T2, int R, int C>
-    inline std::vector<Eigen::Matrix<typename return_type<T1, T2>::type, R, C> >
-    append_array(const std::vector<Eigen::Matrix<T1, R, C> >& x,
-                 const std::vector<Eigen::Matrix<T2, R, C> >& y) {
-      if (!x.empty() && !y.empty()) {
-        check_matching_dims("append_array",
-                            "x[1]",
-                            x.front(),
-                            "y[1]",
-                            y.front());
+    template<typename T1, typename T2>
+    inline void append_array_copy(const std::vector<T1> &x,
+                                  std::vector<T2> &y) {
+      y.resize(x.size());
+      for (size_t i = 0; i < x.size(); i++) {
+        append_array_copy(x[i], y[i]);
       }
+    }
 
-      std::vector<Eigen::Matrix<typename return_type<T1, T2>::type, R, C> > z;
-      z.reserve(x.size() + y.size());
-      for (size_t i = 0; i < x.size(); i++)
-        z.push_back(x[i].template cast<typename return_type<T1, T2>::type>());
-      for (size_t i = 0; i < y.size(); i++)
-        z.push_back(y[i].template cast<typename return_type<T1, T2>::type>());
-      return z;
+    /**
+     * This function checks the sizes of both arguments are consistent.
+     *
+     * The basic version assumes variables are of compatible sizes. There are
+     * specializations for Eigen::Matrices and std::vectors (where this
+     * assumption is not true).
+     *
+     * @tparam T1 Type of x
+     * @tparam T2 Type of y
+     * @param x First argument to check
+     * @param y Second argument to check
+     */
+    template<typename T1, typename T2>
+    inline void append_array_check_size(const T1& x, const T2& y) {
+    }
+
+    /**
+     * This function checks if two Eigen::Matrices are of compatible sizes
+     * by delegating to check_matching_dims.
+     *
+     * @tparam T1 Scalar type of x
+     * @tparam T2 Scalar type of y
+     * @tparam R Eigen RowsAtCompileTime of both matrices
+     * @tparam C Eigen ColsAtCompileTime of both matrices
+     * @param x First matrix
+     * @param y Second matrix
+     */
+    template<typename T1, typename T2, int R, int C>
+    inline void append_array_check_size(const Eigen::Matrix<T1, R, C>& x,
+                     const Eigen::Matrix<T2, R, C>& y) {
+      check_matching_dims("append_array", "x", x, "y", y);
```

Comment:
The `append_array` thing here isn't the name of the function.  Are you sure this only gets called from `append_array`?  If so, add a warning to the doc about the message saying it should only be called from `append_array`.